File: app/src/main/java/com/artmurka/artmurkaapp/Views/Fragments/ItemListFragment.java
package com.artmurka.artmurkaapp.Views.Fragments;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.StaggeredGridLayoutManager;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import com.artmurka.artmurkaapp.Model.Databases.Preferences;
import com.artmurka.artmurkaapp.Model.Pojo.ItemList.ItemList.GoodsProperties;
import com.artmurka.artmurkaapp.Presenter.Adapters.RVitemListAdapterList;
import com.artmurka.artmurkaapp.Presenter.ItemListPresenter;
import com.artmurka.artmurkaapp.Presenter.Adapters.RVitemListAdapter;
import com.artmurka.artmurkaapp.Presenter.InterfacesPresenter.IPresenterItemList;
import com.artmurka.artmurkaapp.R;
import com.artmurka.artmurkaapp.Views.Fragments.Interfaces.IItemListFragment;

import java.util.ArrayList;


public class ItemListFragment extends Fragment implements IItemListFragment {

    private RecyclerView recyclerView;
    private RVitemListAdapter recyclerAdapter;
    private RVitemListAdapterList recyclerAdapterList;
    private String url = "";
    private IPresenterItemList presenter;
    private int curPage = 1;
    private String sort = "name";
    private String order = "asc";
    private int visibleThreshold = 4;
    private Boolean isLoading = false;
    private int totalItemCount, lastVisibleItem;
    private ProgressBar progressBar;

    public ItemListFragment() {
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_item_list, container, false);
        progressBar = (ProgressBar)view.findViewById(R.id.progressBar2);
        Log.d("Log.d", "itemListFragment ");
        Bundle bundle = getArguments();
        if (bundle != null) {
            if (bundle.getString("url") != null) {
                url = bundle.getString("url");
            }
            if (bundle.getString("sort")!=null){
                sort = bundle.getString("sort");
            }
            if (bundle.getString("order")!=null){
                order= bundle.getString("order");
            }
        }
        recyclerView = (RecyclerView) view.findViewById(R.id.recyclerView);
        if (Preferences.getListSettings()==1){

       //     final RecyclerView.LayoutManager recyclerLayoutManager = new GridLayoutManager(view.getContext(), 2);
            StaggeredGridLayoutManager recyclerLayoutManager =new StaggeredGridLayoutManager(2,
                    StaggeredGridLayoutManager.VERTICAL);
            recyclerView.setLayoutManager(recyclerLayoutManager);
            recyclerAdapter = new RVitemListAdapter(view.getContext());
            recyclerView.setAdapter(recyclerAdapter);

            recyclerView.setOnScrollListener(new RecyclerView.OnScrollListener() {
                @Override
                public void onScrollStateChanged(RecyclerView recyclerView, int newState) {
                    super.onScrollStateChanged(recyclerView, newState);
                }

                @Override
                public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                    super.onScrolled(recyclerView, dx, dy);
                    presenter.getCategoriesData(++curPage);
                }
            });

        } else {
            final LinearLayoutManager recyclerLayoutManager = new LinearLayoutManager(view.getContext());
            recyclerLayoutManager.setOrientation(LinearLayout.VERTICAL);
            recyclerView.setLayoutManager(recyclerLayoutManager);
            recyclerAdapterList = new RVitemListAdapterList(view.getContext());
            recyclerView.setAdapter(recyclerAdapterList);

            recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
                @Override
                public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                    super.onScrolled(recyclerView, dx, dy);
                    totalItemCount = recyclerLayoutManager.getItemCount();
                    Log.d("Log.d", totalItemCount + "");
                    lastVisibleItem = ((LinearLayoutManager) recyclerLayoutManager).findLastVisibleItemPosition();
                    Log.d("Log.d", lastVisibleItem + "");
                    if (!isLoading && totalItemCount <= (lastVisibleItem + visibleThreshold)) {
                        presenter.getCategoriesData(++curPage);
                        progressBar.setVisibility(View.VISIBLE);
                        isLoading = true;
                    }
                }
            });
        }



        if (presenter == null) {
            presenter = new ItemListPresenter(this, url, sort, order);
            Preferences.setListUrl(url);
        }
        presenter.getCategoriesData(curPage);
        progressBar.setVisibility(View.VISIBLE);
        return view;
    }

    @Override
    public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true); //for menu
    }


    @Override
    public void showItemList(ArrayList<GoodsProperties> goodsProperties) {
        if (Preferences.getListSettings()==1){
            recyclerAdapter.setData(goodsProperties);
        } else {
            recyclerAdapterList.setData(goodsProperties);
        }
        progressBar.setVisibility(View.INVISIBLE);
        isLoading = false;
    }

    @Override
    public void showError(String error) {
        Snackbar.make(getView(), error, Snackbar.LENGTH_LONG)
                .setAction("Якась помилка...", new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {

                    }
                }).show();
        progressBar.setVisibility(View.INVISIBLE);
    }

    @Override
    public void setTitle(String title) {
        try {
            ((AppCompatActivity) getActivity()).getSupportActionBar().setTitle(title);

        } catch (NullPointerException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void stopProgressBar() {
        progressBar.setVisibility(View.INVISIBLE);
    }


    @Override
    public void onPrepareOptionsMenu(Menu menu) {
        menu.findItem(R.id.sort).setVisible(true);
        super.onPrepareOptionsMenu(menu);
    }
}

File: app/src/main/java/com/artmurka/artmurkaapp/android/views/fragments/ItemListFragment.kt
package com.artmurka.artmurkaapp.android.views.fragments


import android.content.Context
import android.os.Bundle
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.StaggeredGridLayoutManager
import android.util.Log
import android.view.LayoutInflater
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.ProgressBar

import com.artmurka.artmurkaapp.data.model.databases.Preferences
import com.artmurka.artmurkaapp.data.model.pojo.itemlist.itemlist.GoodsProperties
import com.artmurka.artmurkaapp.presenter.adapters.RVitemListAdapterList
import com.artmurka.artmurkaapp.presenter.adapters.RVitemListGridAdapter
import com.artmurka.artmurkaapp.presenter.ItemListPresenter
import com.artmurka.artmurkaapp.presenter.adapters.RVitemListAdapter
import com.artmurka.artmurkaapp.presenter.interfaces_presenter.IPresenterItemList
import com.artmurka.artmurkaapp.R
import com.artmurka.artmurkaapp.android.views.fragments.interfaces.IItemListFragment
import com.artmurka.artmurkaapp.presenter.Presenter
import com.artmurka.artmurkaapp.presenter.PresenterView
import kotlinx.android.synthetic.main.fragment_item_list.*

import java.util.ArrayList
import javax.inject.Inject


class ItemListFragment : BaseFragment(), IItemListFragment, RVitemListAdapter.OnItemClickListener {


    override fun getLayout(): Int = R.layout.fragment_item_list

    override fun getFragmentPresenter(): Presenter<out PresenterView> = presenter
    @Inject
    lateinit var presenter: ItemListPresenter

    private var recyclerAdapter: RVitemListAdapter? = null
    private var recyclerAdapterList: RVitemListAdapterList? = null
    private var recyclerGridAdapter: RVitemListGridAdapter? = null

    private var url: String? = ""
    private var curPage = 1
    private var sort: String? = "name"
    private var order: String? = "asc"
    private val visibleThreshold = 4
    private var isLoading: Boolean? = false
    private var totalItemCount: Int = 0
    private var lastVisibleItem: Int = 0

    override fun toWishList(goodsId: String) = presenter.toWishList(goodsId)

    override fun toBasket(goodId: String) = presenter.toBasket(goodId)

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        presenter.takeView(this)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        Log.d("Log.d", "itemListFragment ")
        val bundle = arguments
        if (bundle != null) {
            if (bundle.getString("url") != null) {
                url = bundle.getString("url")
            }
            if (bundle.getString("sort") != null) {
                sort = bundle.getString("sort")
            }
            if (bundle.getString("order") != null) {
                order = bundle.getString("order")
            }
        }

        if (Preferences.listSettings == 1) {

            val recyclerLayoutManager = GridLayoutManager(view?.context, 2)
            recyclerView.layoutManager = recyclerLayoutManager
            recyclerAdapter = RVitemListAdapter(view?.context!!)
            recyclerAdapter?.clickListener = this
            recyclerView.adapter = recyclerAdapter

            recyclerView.setOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                }

                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    presenter.getCategoriesData(++curPage, url!!, sort!!, order!!)
                }
            })

        } else if (Preferences.listSettings == 2) {
            val recyclerLayoutManager = LinearLayoutManager(view?.context)
            recyclerLayoutManager.orientation = LinearLayout.VERTICAL
            recyclerView.layoutManager = recyclerLayoutManager
            recyclerAdapterList = RVitemListAdapterList(view?.context!!)
            recyclerAdapterList?.clickListener = this
            recyclerView.adapter = recyclerAdapterList


            recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    totalItemCount = recyclerLayoutManager.itemCount
                    Log.d("Log.d", totalItemCount.toString() + "")
                    lastVisibleItem = recyclerLayoutManager.findLastVisibleItemPosition()
                    Log.d("Log.d", lastVisibleItem.toString() + "")
                    if ((!isLoading!!)!! && totalItemCount <= lastVisibleItem + visibleThreshold) {

                        presenter.getCategoriesData(++curPage, url!!, sort!!, order!!)
                        progressBar2.visibility = View.VISIBLE
                        isLoading = true
                    }
                }
            })
        } else if (Preferences.listSettings == 3) {

            val recyclerLayoutManager = StaggeredGridLayoutManager(3,
                    StaggeredGridLayoutManager.VERTICAL)
            recyclerView!!.layoutManager = recyclerLayoutManager
            recyclerGridAdapter = RVitemListGridAdapter(view?.context!!)
            recyclerGridAdapter?.clickListener = this
            recyclerView!!.adapter = recyclerGridAdapter

            recyclerView!!.setOnScrollListener(object : RecyclerView.OnScrollListener() {
                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    super.onScrollStateChanged(recyclerView, newState)
                }

                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    super.onScrolled(recyclerView, dx, dy)
                    presenter.getCategoriesData(++curPage, url!!, sort!!, order!!)
                }
            })
        }

        Preferences.listUrl = this.url!!
        presenter.getCategoriesData(curPage++, url!!, sort!!, order!!)

        progressBar2.visibility = View.VISIBLE
    }


    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true) //for menu
    }


    override fun showItemList(goodsProperties: ArrayList<GoodsProperties>) {
        when (Preferences.listSettings) {
            1 -> recyclerAdapter!!.setData(goodsProperties)
            2 -> recyclerAdapterList!!.setData(goodsProperties)
            3 -> recyclerGridAdapter!!.setData(goodsProperties)
            else -> {
            }
        }
        progressBar2.visibility = View.INVISIBLE
        isLoading = false
    }

    override fun showError(error: String) {
        Snackbar.make(view!!, error, Snackbar.LENGTH_LONG)
                .setAction(resources.getString(R.string.any_error)) { }.show()
        progressBar2.visibility = View.INVISIBLE
    }

    override fun setTitle(title: String) {
        try {
            (activity as AppCompatActivity).supportActionBar!!.title = title

        } catch (e: NullPointerException) {
            e.printStackTrace()
        }

    }

    override fun stopProgressBar() {
        progressBar2.visibility = View.INVISIBLE
    }


    override fun onPrepareOptionsMenu(menu: Menu?) {
        menu?.findItem(R.id.sort)?.isVisible = true
        super.onPrepareOptionsMenu(menu)
    }
}
